File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;                
        int weightSum = 0;
        int idx = 0;
        
        Queue<Integer> bridge = new LinkedList<>();
        while(bridge.size() < bridge_length){
            bridge.add(0);
        }
        
        while(idx < truck_weights.length){
            answer++;
            //System.out.println("weightSum : " + weightSum);
            //System.out.println("bridge : " + bridge);
            
            if(bridge.size() == bridge_length){
                //System.out.println("bridge after size : " + bridge.size());
                weightSum -= bridge.poll();
            }
            
            if(weightSum + truck_weights[idx] <= weight){
                bridge.add(truck_weights[idx]);
                //System.out.println("bridge after : " + bridge);
                weightSum = weightSum + truck_weights[idx];
                idx++;
            } else{
                bridge.add(0);
                //System.out.println("bridge after : " + bridge);
            }                                    
        }
        
        while(weightSum > 0){
            //System.out.println("remain poll : " + bridge);
            weightSum -= bridge.poll();
            answer++;            
        }
        
        //System.out.println(answer);
        
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge_weight = 0;
    bridge_stack = deque()
    
    for _ in range(bridge_length):
        bridge_stack.append(0)
    
    for truck in truck_weights:
        if bridge_weight + truck - bridge_stack[0] <= weight:
            bridge_stack.append(truck)
            bridge_weight += truck
            bridge_weight -= bridge_stack.popleft()
            # print("1 :", bridge_stack)
            answer += 1
        else:
            while bridge_weight + truck - bridge_stack[0] > weight:
                bridge_weight -= bridge_stack.popleft()
                bridge_stack.append(0)
                # print("2 :",bridge_stack)
                answer += 1
            bridge_stack.append(truck)
            bridge_weight += truck
            bridge_weight -= bridge_stack.popleft()
            # print("3 :",bridge_stack)
            answer += 1
        
    while bridge_weight > 0:
        # print("4 :",bridge_stack)
        answer += 1
        bridge_weight -= bridge_stack.popleft()
    
    return answer